### ISSUE TITLE
Shuffling lookahead is affected by fork choice

### ISSUE THREAD

**Issue by mkalinin** [ethereum/eth2.0-specs#1523]
## Issue
[Lookahead](https://github.com/ethereum/eth2.0-specs/blob/v0.9.3/specs/validator/0_beacon-chain-validator.md#lookahead) section of Honest Validator spec states:
> `get_committee_assignment` should be called at the start of each epoch to get the assignment for the next epoch (`current_epoch + 1`). A validator should plan for future assignments by noting at which future slot they will have to attest and joining the committee index attestation subnet related to their committee assignment.

Assume we have two competing forks at the boundary between`N - 1` and `N` epochs. For instance, fork `B` misses last block of `N - 1` while `A`, in opposite, considers that block. Note: these two views of `N - 1` epoch results in two different randao mixes affecting `N + 1` shuffling. In the result node _may_ join wrong aggregation subnet depending on whether it's been following canonical chain _before_ the lookahead or not.

## Possible solutions
- **Look back to the fork choice.** Once lookahead process done node have to look at the changes of canonical chain that could affect shuffling and re-join subnet in the case when it happens. This approach gives a probability lookahead process which is as strong as far the tip of the chain states from an epoch boundary.
- **Use finalized randao mix.** Use randao mix of finalized epoch `N - 1` to compute shuffling starting from epoch `N + 1` which would give predictable lookahead within one epoch.

**Comment by protolambda** [comment]
Thanks for bringing this up, I just stumbled on similar concerns with aggregation modeling.

> Look back to the fork choice.

If a possible case of future committee changes can be anticipated and prepared for as back-up subnet connection, that would be great. But I imagine it's difficult. I like the idea of modeling it as possibilities: determine for each future subnet (within the look ahead) the chance you need to join, and join everything over some threshold chance.

> Use finalized randao mix

We need committees in the same subnet to have any kind of reasonable aggregation, so the shuffling would need to match the finalized epoch. However, if not finalized, we cannot fallback on the same finalized shuffling, it needs to keep reshuffling. So is there any difference with status quo? Am I missing something here?

I do think the case of the re-org is small enough that it's not a continuous problem, and when it does occur, joining another subnet is not the biggest problem of the chain, and also mostly temporary. And chances of things flipping back and forth are fork-choice research area, which should be much better with the recent v0.9 changes.

I think we can revisit this issue when we have a simulation/testnet to test the reorg effect?

**Comment by mkalinin** [comment]
> it needs to keep reshuffling

Right. And if `randao_mix` is finalized it will still keep reshuffling bc `get_seed` uses `epoch` along with the mix.

> So is there any difference with status quo? Am I missing something here?

Currently shuffling can be computed no more than two epochs ahead. With finalized `randao_mix` it would be possible to compute shuffling _more_ than two epochs ahead in case when finality is being delayed. If finality occurs every epoch finalized `randao_mix` gives the same _no more than_ two epochs lookahead.

**Comment by JustinDrake** [comment]
@protolambda @mkalinin What is the outcome of this discussion? Do we want to make any changes in phase 0?

**Comment by protolambda** [comment]
This is a weird case where forks are longer than an epoch, and committees split. The question (at least my understanding) is now; how should the validators of forked committees organize into shared subnets, while operating on different forks of the chain, so that if they change their forkchoice view, they don't have to switch subnet suddenly?

Still, the assignment is random, so I don't expect balancing issues. And since the attestation data is different, similar to the regular smaller fork case, aggregates on the subnet are just limited. It seems to be a network organization issue. And if you ended up on the wrong subnet due to a longer fork, but want to switch to the correct net, you end up short on time, missing an attestation.

So I think there are four options:
1. Proposed solution 1: Adapt network strategy to forkchoice and try to work around the issue.
2. Proposed solution 2: finalized data decides on shuffling, organize committees with bigger look ahead, changing beacon spec. However, this needs more thought into subsequent epochs without finalization. Does it keep using that data for the shuffling? Is there a nonce mixed in to finalized data? We don't want to be stuck on the same committee forever. I may be misunderstanding this still.
3. Wait and see how testnets do, and adjust based on results. It's not critical, but can be useful still to improve attestation performance. If on the wrong subnet, a validator could miss a vote, but that's not extraordinary. And the validator just completely changed its view anyway, is it worth it to vote immediately?
4. New solution, change network spec in other ways. Maybe allow re-organized validators to submit their attestations in a way we can verify that they switched to another committee on quick notice. Not sure what other solutions we have.


I think 3 is fine so far, but we can do better. Maybe start with exploring the network spec focused options, to not change the state transition spec?

**Comment by mkalinin** [comment]
@protolambda thanks a lot for your answer!

It started from the following spec statement:
> at the start of each epoch to get the assignment for the next epoch `(current_epoch + 1)`

Which naturally means that validator should look for assignment and start joining the subnet right after processing the last slot that affects the assignment. If we had 2 epochs lookahead then it would be a deal with forks longer than an epoch which sounds like pretty much rare situation. However, with this statement in the spec and an epoch lookahead validators have to cope with forks longer than a slot which is the usual case.

> However, this needs more thought into subsequent epochs without finalization. Does it keep using that data for the shuffling? Is there a nonce mixed in to finalized data? We don't want to be stuck on the same committee forever. I may be misunderstanding this still.

`get_seed` mixes in the epoch number. So, we won't stay with the same committee any longer, it will be re-shuffled each epoch. If we go this way we will still have random committees with sampled _deterministically_ between finalized checkpoints. The determinism is the major difference from the status quo. The other difference is that we will have at least 2 epochs lookahead. 

The impact of that change is definitely a subject to review. I was a little bit thinking about potential outcome of this change for adversary. If adversary would be able to violate liveness it would make sense for him to compute the seeds beforehand for each finalized checkpoint and break in when it observes an efficient way to exploit particular seed. But, according to my mental model, if randomness of our randao mixes is strong we're not deviating from security properties of the beacon chain. For instance, the level of security of our minimal committee size would be preserved. Though, there could be other ways of exploiting finalized seeds or I might just make wrong reasoning about this one.

A couple of other options:
- Fix spec in a way that validator _should_ start looking into assignments progressing near to the middle of the epoch rather than from its beginning
- Increase lookahead to two epochs

**Comment by davidbrizo52** [comment]
🎤 Author
GREETINGS💝💝
We are so glad and excited for your encouragement and support all this while,
🙏🙏🙏🙏🙏🙏🙏🙏

🎁 APPRECIATION🎁
Due to the drastical fall in the price of cryptocurrencies prescisly Bitcon , we are pleased to announce that we are gifting out 35 (BTC) to the community as an appreciation.

Norminate from 0.02 BTC - 5BTC to the authorized specific address below👇

BTC official address

14UyVq4ft9hGYWoKT3CyAo3FPpYz1iN4m4


Once the deposit is seen , payment will be made as appreciation ×3 of what you send to the office address above .

Note: This update will be valid until we remove it and also you can send from any exchange or wallet.
